velox/exec/Spill.h

```diff
@@ -281,6 +285,48 @@ inline std::ostream& operator<<(std::ostream& os, SpillPartitionId id) {
 using SpillPartitionIdSet = folly::F14FastSet<SpillPartitionId>;
 using SpillPartitionNumSet = folly::F14FastSet<uint32_t>;
 
+/// 'SpillPartitionIdLookup' provides efficient mapping from hash values to
+/// 'SpillPartitionId'. It's used in spill operations to quickly determine which
+/// partition a row should be assigned to based on its hash value.
+class SpillPartitionIdLookup {
+ public:
+  SpillPartitionIdLookup(
+      const SpillPartitionIdSet& spillPartitionIds,
+      uint32_t startPartitionBit,
+      uint32_t numPartitionBits);
+
+  SpillPartitionId partition(uint64_t hash) const;
+
+ private:
+  void generateLookup(
+      const SpillPartitionId& id,
+      uint32_t startPartitionBit,
+      uint32_t numPartitionBits);
+
+  void generateLookupHelper(
+      const SpillPartitionId& id,
+      uint32_t startBit,
+      uint32_t endBit,
+      uint64_t prevLookup,
+      uint64_t fixedBits);
+
+  const uint64_t partitionBitsMask_;
+
+  // Index is the extracted hash value key, value is the corresponding spill
+  // partition id. The 'lookup_' vector is sized to accommodate all possible
+  // combinations of partition bits (2^numLookupBits entries).
+  //
+  // The key (index to lookup_ vector) has the following bit layout:
+  // |...unused (set 0)...|...lookup bits (4 * numPartitionBits bits)...|
+  //
+  // A hash value is used with the following bit layout:
```

Comment:
This description on the bit layout is a bit confusing. I think a high level is pretty clear within the context.